### PR TITLE
This changes the way we do add/sub/compare operators 

### LINKED ISSUE #158: Value comparisons for infinity have different behavior from dimensionless floats
```python
>>> from labrad.units import s
>>> float('inf') == float('inf')
True
>>> float('inf')*s == float('inf')*s
False
```

This is due to the fact that \_\_cmp\_\_ for WithUnit is implemented as
```python
def __cmp__(self, other):
        '''
        This just returns the difference between self and other.
        We don't call cmp() on it to convert to +/-1 or 0 because
        that doesn't work on complex numbers, even if we only want
        to test for equality.
        '''
        return self._sum(other, 1, -1)
```
and inf - inf = nan

### PR DESCRIPTION
**Pull request by ejeffrey**
fixes #158.

This gets rid of the old __cmp__ based comparison operators.

### CODE REVIEW

File: labrad/units.py [hunk 1]
```diff
@@ -268,63 +268,71 @@ def units(self):
...
         elif self.is_dimensionless:
-            return sign1 * self._value + sign2 * other / self.unit.conversionFactorTo('')
+            return (self[''], other, U.Unit(''))
         elif other == 0:
-            return sign1 * self._value
+            return (self._value, 0.0, self.unit)
         else:
             raise TypeError("Incompatible Units %s, ''" % (self.unit,))
-
+
```
Comment (DanielSank) [inline labrad/units.py]: Remote spaces.

File: labrad/units.py [hunk 2]
```diff
@@ -268,63 +268,71 @@ def units(self):
         """A string representation of the unit of this value."""
         return self.unit.name
 
-    def _sum(self, other, sign1, sign2):
+    def _convert_units(self, other):
+        """Convert two objects to use compatible units.
```
Comment (DanielSank) [inline labrad/units.py]: Might want to point out in the doc that we return `Unit('')` if self is dimensionless.

File: labrad/units.py [hunk 3]
```diff
@@ -268,63 +268,71 @@ def units(self):
...
-        We don't call cmp() on it to convert to +/-1 or 0 because
-        that doesn't work on complex numbers, even if we only want
-        to test for equality.
-        '''
-        return self._sum(other, 1, -1)
-
     def __lt__(self, other):
-        return self.__cmp__(other) < 0
+        a, b, unit = self._convert_units(other)
+        return a<b
```
Comment (DanielSank) [inline labrad/units.py]: White space here and in the next three methods.

**Comment by DanielSank** [comment]
LGTM. Do you have any notion as to whether these changes will cause any performance issues?

**Comment by ejeffrey** [comment]
I don't think it will make any difference (positive or negative).  The sequence of steps and number of function calls is essentially the same, just with calls to convert_units instead of sum.  It would be nice to have some standard performance tests.

**Comment by DanielSank** [comment]
Good idea. What was causing the problems in the past? Was it unit conversion, comparisons,...?

**Comment by ejeffrey** [comment]
I think it was just value multiplication which requires constructing new Unit and Value objects for each multiplication.  There are a lot of function calls for each operation, and function call overhead in python is pretty high.  There was no smoking gun single point that was taking all the time.

**Pull request merged**